File: cpp/cpp_100-999/565.cc
// C++
// leetcode 565
// https://leetcode.cn/problems/array-nesting/

class Solution {
public:
    int arrayNesting(vector<int>& nums) {
    	int n = nums.size(), res = 0;
        for (int i = 0; i < n; i++) {
            int cur = i, cnt = 0;
            while (nums[cur] != -1) {
                cnt++;
                int c = cur;
                cur = nums[cur];
                nums[c] = -1;
            }
            res = max(res, cnt);
        }
    	return res;
    }
};

File: Go/Go_100-999/565.go
// Go
// leetcode 565
// https://leetcode.cn/problems/array-nesting/

func arrayNesting(nums []int) int {
	n, res := len(nums), 0
	for i := 0; i < n; i++ {
		cur, cnt, tmp := i, 0, 0
		for nums[cur] != -1 {
			cnt++
			tmp = cur
			cur = nums[cur]
			nums[tmp] = -1
		}
		if cnt > res {
			res = cnt
		}
	}
	return res
}